pkg/controllers/monitoring-stack/controller.go

```diff
@@ -0,0 +1,179 @@
+type Options struct {
+	InstanceSelector string
+}
+
+// RBAC for managing monitoring stacks
+//+kubebuilder:rbac:groups=monitoring.rhobs,resources=monitoringstacks,verbs=list;watch;create;update
+//+kubebuilder:rbac:groups=monitoring.rhobs,resources=monitoringstacks/status,verbs=get;update
+
+// RBAC for managing Prometheus CRs
+//+kubebuilder:rbac:groups=monitoring.coreos.com,resources=prometheuses;servicemonitors,verbs=list;watch;create;update;delete
+//+kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=roles;rolebindings,verbs=list;watch;create;update;delete
+//+kubebuilder:rbac:groups="",resources=serviceaccounts;secrets,verbs=list;watch;create;update;delete
+
+// RBAC for managing Grafana CRs
+//+kubebuilder:rbac:groups=integreatly.org,namespace=monitoring-stack-operator,resources=grafanadatasources,verbs=list;watch;create;update;delete
+
+// RBAC for delegating permissions to Prometheus
+//+kubebuilder:rbac:groups="",resources=pods;services;endpoints,verbs=get;list;watch
+//+kubebuilder:rbac:groups=extensions;networking.k8s.io,resources=ingresses,verbs=get;list;watch
+
+// RegisterWithManager registers the controller with Manager
+func RegisterWithManager(mgr ctrl.Manager, opts Options) error {
+	split := strings.Split(opts.InstanceSelector, "=")
+	if len(split) != 2 {
+		return fmt.Errorf("invalid InstanceSelector: %s", opts.InstanceSelector)
+	}
+
+	r := &reconciler{
+		k8sClient:             mgr.GetClient(),
+		scheme:                mgr.GetScheme(),
+		logger:                ctrl.Log.WithName("monitoring-stack-operator"),
+		instanceSelectorKey:   split[0],
+		instanceSelectorValue: split[1],
+	}
+
+	// We only want to trigger a reconciliation when the generation
+	// of a child changes. Until we need to update our the status for our own objects,
+	// we can save CPU cycles by avoiding reconciliations triggered by
+	// child status changes.
+	p := predicate.GenerationChangedPredicate{}
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&stack.MonitoringStack{}).
+		Owns(&monv1.Prometheus{}).WithEventFilter(p).
+		Owns(&v1.ServiceAccount{}).WithEventFilter(p).
+		Owns(&rbacv1.Role{}).WithEventFilter(p).
+		Owns(&rbacv1.RoleBinding{}).WithEventFilter(p).
+		Owns(&monv1.ServiceMonitor{}).WithEventFilter(p).
+		Complete(r)
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.logger.WithValues("stack", req.NamespacedName)
+	logger.Info("Reconciling monitoring stack")
+	ms, err := r.getStack(ctx, req)
+	if err != nil {
+		// retry since some error has occured
+		return ctrl.Result{}, err
+	}
+	if ms == nil {
+		// no such monitoring stack, so stop here
+		return ctrl.Result{}, nil
+	}
+
+	patchers, err := r.stackComponentPatchers(ms, r.instanceSelectorKey, r.instanceSelectorValue)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for _, mapper := range patchers {
+		if err := r.reconcileObject(ctx, ms, mapper, logger); err != nil {
```

Comment:
Just a thought ... will this do a partial reconciliation and leave system from a stable to unstable state?

Reply:
Yes, I think it could. I am just not sure how to solve it.